pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,282 @@
+		return nil, err
+	}
+
+	return &migrator{
+		client:           dynamic,
+		discoveryClient:  clientset.Discovery(),
+		crdClient:        crd.ApiextensionsV1().CustomResourceDefinitions(),
+		apiserviceClient: apiservice.ApiregistrationV1().APIServices(),
+	}, nil
+}
+
+func (d *migrator) Start(ctx context.Context) (*MigrationResultList, error) {
+	return d.start(ctx)
+}
+
+func (d *migrator) start(ctx context.Context) (*MigrationResultList, error) {
+	schemas, err := d.findMigratableResources(ctx)
+	if err != nil {
+		return nil, err
+	}
+	results := &MigrationResultList{
+		Status: MigrationSuccess,
+	}
+	errorOccured := false
+	start := time.Now()
+	klog.Info("schema migration started")
+
+	wg := sync.WaitGroup{}
+	lock := sync.Mutex{}
+	wg.Add(len(schemas))
+	for _, sch := range schemas {
+		go func(wg *sync.WaitGroup, sch schema.GroupVersionResource) {
+			defer wg.Done()
+			objectList := &unstructured.UnstructuredList{}
+			var migrationErr error
+
+			migrationErr = retry.OnError(retry.DefaultBackoff, canRetry, func() error {
+				objectList, migrationErr = d.list(ctx, sch, metav1.ListOptions{})
+				if migrationErr != nil {
+					return migrationErr
+				}
+				return nil
+			})
+
+			if migrationErr != nil {
+				errorOccured = true
+				migrationErr = fmt.Errorf("could not list resources: %+v", migrationErr)
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{Error: migrationErr, ResourceVersion: sch, Timestamp: time.Now()})
+				lock.Unlock()
+				return
+			}
+
+			for _, object := range objectList.Items {
+				ref := object
+				migrationErr := d.migrateOneItem(ctx, sch, &ref)
+				if migrationErr != nil {
+					errorOccured = true
+				}
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{
+					Error:           migrationErr,
+					ResourceVersion: sch,
+					ObjectMeta:      getObjectMeta(&ref),
+					Timestamp:       time.Now()})
+				lock.Unlock()
+			}
+		}(&wg, sch)
+	}
+	if errorOccured {
+		results.Status = MigrationFailure
+	}
+	wg.Wait()
+	klog.Infof("schema migration finished, it took %s to complete", time.Since(start).String())
+	return results, nil
+}
+
+func (d *migrator) findMigratableResources(ctx context.Context) ([]schema.GroupVersionResource, error) {
+	customGroups, err := d.findCustomGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+	aggregatedGroups, err := d.findAggregatedGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resourceToGroupVersions := make(map[string][]schema.GroupVersion)
+	_, resourceLists, err := d.discoveryClient.ServerGroupsAndResources()
+	if err != nil {
+		return nil, err
+	}
+	for _, resourceList := range resourceLists {
+		gv, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+		if err != nil {
+			klog.Errorf("cannot parse group version %s, ignored", resourceList.GroupVersion)
+			continue
+		}
+		if customGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's a custom group", gv.Group)
```

Comment:
I'd be careful with `V(4)`. Currently we only show `0` level. We might (at least for some time) have them on level 0 and see how many logs this produces

Reply:
Sure thing, updated